features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c

```diff
@@ -0,0 +1,333 @@
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    if( cmac_ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
+
+    if( ( ret = SaSi_AesFinish( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_len,
+                                cmac_ctx->unprocessed_block,
+                                cmac_ctx->unprocessed_len, output, &olen ) ) != 0 )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+exit:
+    if( SaSi_AesFree( &cmac_ctx->CC_Context ) != 0 && ret == 0 )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+    }
+
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
+
+    if( ctx == NULL || ctx->cipher_info == NULL || ctx->cmac_ctx == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    /* Reset the internal state */
+    cmac_ctx->unprocessed_len = 0;
+    mbedtls_platform_zeroize( cmac_ctx->unprocessed_block,
+                              sizeof( cmac_ctx->unprocessed_block ) );
+
+    return( 0 );
+}
+
+/**
+ * \brief               This function calculates the full generic CMAC
+ *                      on the input buffer with the provided key.
+ *
+ *                      The function allocates the context, performs the
+ *                      calculation, and frees the context.
+ *
+ *                      The CMAC result is calculated as
+ *                      output = generic CMAC(cmac key, input buffer).
+ *
+ *
+ * \param cipher_info   The cipher information.
+ * \param key           The CMAC key.
+ * \param keylen        The length of the CMAC key in bits.
+ * \param input         The buffer holding the input data.
+ * \param ilen          The length of the input data.
+ * \param output        The buffer for the generic CMAC result.
+ *
+ * \return              \c 0 on success.
+ * \return              #MBEDTLS_ERR_MD_BAD_INPUT_DATA
+ *                      if parameter verification fails.
+ */
+int mbedtls_cipher_cmac( const mbedtls_cipher_info_t *cipher_info,
+                         const unsigned char *key, size_t keylen,
+                         const unsigned char *input, size_t ilen,
+                         unsigned char *output )
+{
+    int ret = 0;
+    mbedtls_cipher_context_t ctx;
+    size_t olen = SASI_AES_BLOCK_SIZE_IN_BYTES;
+
+    if( cipher_info == NULL || key == NULL ||
+        input == NULL || output == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    mbedtls_cipher_init( &ctx );
+
+    if( ( ret = mbedtls_cipher_setup( &ctx, cipher_info ) ) != 0 )
+        goto exit;
+
+    ret = mbedtls_cipher_cmac_starts( &ctx, key, keylen );
+    if( ret != 0 )
+        goto exit;
+
+    if( ( ret = init_cc( ctx.cmac_ctx ) ) != 0 )
+    {
+        goto exit;
+    }
+
+    if( SaSi_AesFinish( &ctx.cmac_ctx->CC_Context, ilen, ( uint8_t * ) input,
```

Comment:
Why not call `mbedtls_cipher_cmac_update()` and `mbedtls_cipher_cmac_finish()?

If AES_ALT is set, won't the software CMAC implementation use it?

Reply:
It can be done, but it adds additional redundant code. In this API, it's a one call CMAC functionality, without storing `unprocessed_block`, and call directly to the CC AES driver to calculate the CMAC.
The alternative implementation of `mbedtls_cipher_cmac_update()` and `mbedtls_cipher_cmac_finish()` has support for the unprocessed data, to support sequential calls with sizes that are not multiple of AES  block size

Reply:
OK